### ISSUE TITLE
Unable to authenticate over ssh with ecdsa keys in OpenSSH format

### ISSUE THREAD

**Issue by donoghuc** [puppetlabs/bolt#1165]
## Describe the Bug
SSH 7.8 switched ssh-keygen to outputting keys in the OpenSSH format (rather than OpenSSL's PEM format). The underlying `net-ssh` gem only supports ed25519 and rsa keys in the OpenSSH format. That issue is tracked with https://github.com/net-ssh/net-ssh/issues/657

Example error:
```
[root@ebo9k1qf1fpcs67 Boltdir]# bolt command run whoami -n ecdsa-openssh
Started on localhost...
Failed on localhost:
Failed to connect to localhost: the given identity is known, but the private key could not be loaded: Net::SSH::Exception (Cannot decode private key of type ecdsa-sha2-nistp256)
Failed on 1 node: localhost
Ran on 1 node in 0.16 seconds
```

## Expected Behavior
Private keys with the new OpenSSH format should be supported with net-ssh and bolt in addition to the `pem` format that is already supported.

**Comment by donoghuc** [comment]
We should add this as a known issue, given the bug report is tracked in net-ssh.

**Comment by davidalexander83** [comment]
Major issue when showcasing cross-platform support in the Bolt workshops, obviously. Would love to see the resolution to this.

**Comment by renich** [comment]
This solved my problem: `gem install ed25519 bcrypt_pbkdf`

```
[renich@introdesk test]$ bolt command run 'echo $HOSTNAME' --user=root --targets all
CLI arguments ["user"] may be overridden by Inventory: /home/renich/Desktop/bolt/test/inventory.yaml
Started on web1.woralelandia.com...
Started on web2.woralelandia.com...
Started on web3.woralelandia.com...
Started on web4.woralelandia.com...
Started on introlap...
Started on introdesk...
Failed on introdesk:
  Sudo password for user renich was not provided for localhost
Failed on web2.woralelandia.com:
  unsupported key type `ssh-ed25519'
  net-ssh requires the following gems for ed25519 support:
   * ed25519 (>= 1.2, < 2.0)
   * bcrypt_pbkdf (>= 1.0, < 2.0)
  See https://github.com/net-ssh/net-ssh/issues/565 for more information
  Gem::MissingSpecError : "Could not find 'ed25519' (~> 1.2) among 234 total gem(s)
  Checked in 'GEM_PATH=/home/renich/.gem/ruby:/usr/share/gems:/usr/local/share/gems', execute `gem env` for more information"
Failed on web3.woralelandia.com:
  unsupported key type `ssh-ed25519'
  net-ssh requires the following gems for ed25519 support:
   * ed25519 (>= 1.2, < 2.0)
   * bcrypt_pbkdf (>= 1.0, < 2.0)
  See https://github.com/net-ssh/net-ssh/issues/565 for more information
  Gem::MissingSpecError : "Could not find 'ed25519' (~> 1.2) among 234 total gem(s)
  Checked in 'GEM_PATH=/home/renich/.gem/ruby:/usr/share/gems:/usr/local/share/gems', execute `gem env` for more information"
Failed on web4.woralelandia.com:
  unsupported key type `ssh-ed25519'
  net-ssh requires the following gems for ed25519 support:
   * ed25519 (>= 1.2, < 2.0)
   * bcrypt_pbkdf (>= 1.0, < 2.0)
  See https://github.com/net-ssh/net-ssh/issues/565 for more information
  Gem::MissingSpecError : "Could not find 'ed25519' (~> 1.2) among 234 total gem(s)
  Checked in 'GEM_PATH=/home/renich/.gem/ruby:/usr/share/gems:/usr/local/share/gems', execute `gem env` for more information"
Failed on web1.woralelandia.com:
  unsupported key type `ssh-ed25519' 
  net-ssh requires the following gems for ed25519 support:
   * ed25519 (>= 1.2, < 2.0)
   * bcrypt_pbkdf (>= 1.0, < 2.0)
  See https://github.com/net-ssh/net-ssh/issues/565 for more information
  Gem::MissingSpecError : "Could not find 'ed25519' (~> 1.2) among 234 total gem(s)
  Checked in 'GEM_PATH=/home/renich/.gem/ruby:/usr/share/gems:/usr/local/share/gems', execute `gem env` for more information"
Failed on introlap:
  Failed to connect to introlap: No route to host - connect(2) for 192.168.86.26:22
Failed on 6 targets: web1.woralelandia.com,web2.woralelandia.com,web3.woralelandia.com,web4.woralelandia.com,introlap,introdesk
Ran on 6 targets in 3.22 sec

[renich@introdesk test]$ gem install ed25519 bcrypt_pbkdf
Fetching ed25519-1.2.4.gem
Building native extensions. This could take a while...
Successfully installed ed25519-1.2.4
Fetching bcrypt_pbkdf-1.0.1.gem
Building native extensions. This could take a while...
Successfully installed bcrypt_pbkdf-1.0.1
2 gems installed
  
[renich@introdesk test]$ bolt command run 'echo $HOSTNAME' --user=root --targets all
CLI arguments ["user"] may be overridden by Inventory: /home/renich/Desktop/bolt/test/inventory.yaml
Started on web2.woralelandia.com...
Started on web3.woralelandia.com...
Started on web1.woralelandia.com...
Started on web4.woralelandia.com...
Started on introlap...
Started on introdesk...
Failed on introdesk:
  Sudo password for user renich was not provided for localhost
Failed on introlap:
  Failed to connect to introlap: No route to host - connect(2) for 192.168.86.26:22
Finished on web2.woralelandia.com:
  STDOUT: 
    woralelandia 
Finished on web3.woralelandia.com:
  STDOUT:
    web3-woralelandia-com
Finished on web1.woralelandia.com:
  STDOUT:
    web1.woralelandia.com
Finished on web4.woralelandia.com:
  STDOUT:
    web4.woralelandia.com
Successful on 4 targets: web1.woralelandia.com,web2.woralelandia.com,web3.woralelandia.com,web4.woralelandia.com
Failed on 2 targets: introlap,introdesk
Ran on 6 targets in 30.71 sec

```

**Comment by donoghuc** [comment]
We ship ed25591 1.2.4 and bcrypt_pbkdf 1.0.0 with system packages. Couple questions:
1. did you install bolt from system packages? Or as a gem
2. Are you sure the ecdsa key is in openssh format?

My suspicion is that perhaps you had installed bolt as a gem (and we do not require those net-ssh extension gems in our gemspec) and that your key is in the Openssl PEM format. When you installed the other dependencies we ship in system packages your ecdsa key worked as expected. If you could confirm that your key is in the newer openssh format please let us know and we can look in to re-evaluating if this is still a problem.

**Comment by renich** [comment]
@donoghuc I installed as a gem.

**Comment by donoghuc** [comment]
ed25519 is supported. ecdsa-sha2-nistp256 is not https://github.com/net-ssh/net-ssh/issues/657#issuecomment-462662686

**Comment by renich** [comment]
Ah! My bad then! ;p

**Comment by donoghuc** [comment]
Was hoping it was magically fixed! Thanks for the input, we realized looking in to this that there is a newer bcrypt_pbkdf we can ship :)

**Comment by lucywyman** [comment]
We recently added a config option that allows you to shell out to SSH instead of using the net-ssh library, which is useful for exactly this case. The last thing to do is document using the option to use an algorithm that isn't supported by net-ssh, which we have an open PR for https://github.com/puppetlabs/bolt/pull/1892  Going to close this as a duplicate of #733 since this is now the same issue that's keeping that issue open, and should also be solved by #1892
